### PR TITLE
Add option to require qualification for registered events.

### PR DESCRIPTION
**Pull request by timreyn**
Adds an option to the competition form to require that competitors be qualified for events that they are registering for.

This does not prevent registering for events with a "Top N" qualification, because it's not yet known what the qualifying standard will be. In theory, we could do "Top N of people currently registered", but that would lead to weird effects if someone qualified cancelled their registration.

Part of #2051.

### CODE REVIEW

**Comment by timreyn** [comment]
Here's what this looks like when you are not qualified for an event:

image

File: WcaOnRails/app/views/registrations/_register_form.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/registrations/_register_form.html.erb:4]: Nit: `.map(&:event)`

Comment (gregorbg) [inline WcaOnRails/app/views/registrations/_register_form.html.erb:4]: More nit: The lambda parameter in `select` should be called `ce` to avoid confusing `competition_event` and `event`.

File: WcaOnRails/app/views/registrations/edit.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/registrations/edit.html.erb:16]: Same two nits as above. Why is this code duplicated anyways?

File: WcaOnRails/app/views/shared/_associated_events_picker.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/shared/_associated_events_picker.html.erb:51]: I had to pause and ponder for several seconds why you're introducing a `hidden_field` for `_destroy` in addition to the checkbox above. Then I realised that the if condition here and the disabled condition are mutually exclusive. On that note, wouldn't it make more sense to do something along the lines of
```ruby
<% if e_disabled %>
  <%= f.hidden_field "_destroy", { value: "1" } %>
<% else %>
  <%= f.check_box "_destroy", ....... %>
<% end %>
```

File: WcaOnRails/config/locales/en.yml
Comment (gregorbg) [inline WcaOnRails/config/locales/en.yml:346]: The text here doesn't seem to fit the description. Where is this used?
(I'd expect something more along the lines of "Registering without automated checking of qualification" or "Registering without being qualified is allowed")

File: WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb
Comment (gregorbg) [inline WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb:3]: Your `bin/rails` runner seems to still generate migrations on version 5.2.
When did you run `bundle install` for the last time? 😅

File: WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb
Comment (gregorbg) [inline WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb:5]: What's the rationale behind using `true` as default?

File: WcaOnRails/lib/qualification.rb
Comment (gregorbg) [inline WcaOnRails/lib/qualification.rb:45]: Ruby syntactic sugar: `return false if user.person.nil?`

Comment (gregorbg) [reply WcaOnRails/lib/qualification.rb:45]: (Yes, as a one-liner.)

File: WcaOnRails/lib/qualification.rb
Comment (gregorbg) [inline WcaOnRails/lib/qualification.rb:64]: Not a fan of self-overriding variables. Call the first occurence `before_deadline_results` or something?

File: WcaOnRails/lib/qualification.rb
Comment (gregorbg) [inline WcaOnRails/lib/qualification.rb:42]: Debating with myself whether this should be split up into `qualifying_results_for(user, event_id)` here and the actual `can_register?` over in `competition_event`. I have no strong arguments other than gut feeling. What is your opinion?

File: WcaOnRails/app/views/shared/_associated_events_picker.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/shared/_associated_events_picker.html.erb:48]: Can we have the icon appear in some gentle tone of red (along the lines of `#FF7F7F`) when you're not qualified? Just to distinguish between "It's not selected because you haven't clicked it yet" and "It's not selected because you _cannot_ click it"

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thanks, very succinct implementation in general! Cosmetic remarks here and there, though.

File: WcaOnRails/lib/qualification.rb
Comment (timreyn) [reply WcaOnRails/lib/qualification.rb:42]: The tricky thing here is "ranking" qualifications. If we did that, the code in `qualification.rb` would either need to ignore "ranking", or be more complicated to consider registered competitors (and then have `CompetitionEvent.can_register?` ignore it). I went with this naming scheme to accurately represent what the code actually checks: whether, based on the qualification standard, the competitor is allowed to register for the event.

For what it's worth, the code to remove unqualified competitors won't use this; it'll do one big SQL query to generate the psych sheet as of the deadline, to avoid doing N SQL queries for N competitors.

File: WcaOnRails/lib/qualification.rb
Comment (timreyn) [reply WcaOnRails/lib/qualification.rb:45]: Thank you!

File: WcaOnRails/lib/qualification.rb
Comment (timreyn) [reply WcaOnRails/lib/qualification.rb:64]: Done.

File: WcaOnRails/app/models/result.rb
Comment (timreyn) [reply WcaOnRails/app/models/result.rb:46]: I'm not sure that applies, because that's about solves in a multi-day round, rather than a multi-day competition.

The really correct way to do this would be based on the competition schedule. That sounds like a more complicated join, though, and I'm not that enthusiastic about trying to write that join.

Maybe this is a question for WCAT? I'll start a thread with them. WCRP does not specify:

> 5.1 The competitor limit can be on either a “first come first serve” basis or based upon
qualification results or rankings in the WCA world rankings of a previously announced
calendar date.

Comment (timreyn) [reply WcaOnRails/app/models/result.rb:46]: I changed this to `:before` and used `end_date` rather than `start_date`, and strictly-less-than.

**Comment by gregorbg** [comment]
Thanks for the changes, code is looking good in its current state but there are a few interesting test fails.
Also, what about the Ranking qualification SQL check that you mentioned in one of the comments? Do you plan on adding that as part of this PR?

**Comment by timreyn** [comment]
Looking into the test failures (but I'm a little short on time this weekend due to Euros! Probably the next round of changes will come early next week).

Re the SQL query -- this is coming in my next PR, which adds a button to remove non-qualified competitors from events (see https://github.com/thewca/worldcubeassociation.org/compare/master...timreyn:worldcubeassociation.org:qualification_clear?expand=1 if you're interested, but it's not ready for an actual review yet).

File: WcaOnRails/app/models/registration.rb
Comment (gregorbg) [reply WcaOnRails/app/models/registration.rb:239]: In contrast to #7050 (which has a very strong motivation because the WCIF PATCH endpoint is even exposed to the OAuth public!!), deleting events for which competitors have already registered can sometimes happen. So your code should be able to deal with it anyways.

But interesting how `registration_competition_events` does not seem to be updated when overall events are deleted... Maybe this will actually be implicitly fixed by #7050, who knows :thinking: 

In terms of this change, what about `!event.competition_event&.can:register?(user)`? The `&` is a nil-safe operator in Ruby, and if the event doesn't exist it breaks out of the call chain and returns nil which then gets negated to `true` so it ultimately gets rejected.

Comment (timreyn) [reply WcaOnRails/app/models/registration.rb]: Ah thank you!

**Comment by timreyn** [comment]
Friendly ping @gregorbg -- I think this should be good to go, pending another rebase!

File: WcaOnRails/app/views/shared/_associated_events_picker.html.erb
Comment (gregorbg) [reply WcaOnRails/app/views/shared/_associated_events_picker.html.erb]: Fine by me!

**Comment by gregorbg** [comment]
Thanks, sorry this has slipped my attention. Going through the remaining open conversations now :)

File: WcaOnRails/app/views/registrations/edit.html.erb
Comment (gregorbg) [reply WcaOnRails/app/views/registrations/edit.html.erb]: That does make sense, but it shouldn't keep you from refactoring the common parts of the code into some helper method that can be invoked on both separate pages :)

File: WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb
Comment (gregorbg) [reply WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb]: I think you should be able to do this as part of the current migration. Something along the lines of
```ruby
def change
  # N.B. This has the default set to false
  add_column ...foo... default: false
  
  # Except all of the competitions that are currently open
  Competition.where("registration_close >= :close", close: DateTime.now).update_all(allow_registration_without_qualification: true)
end
```

Comment (timreyn) [reply WcaOnRails/db/migrate/20220627195217_add_allow_registration_without_qualification_to_competitions.rb]: Done!

File: WcaOnRails/app/views/registrations/edit.html.erb
Comment (timreyn) [reply WcaOnRails/app/views/registrations/edit.html.erb]: Ah, got it :) Added a method to Competition.

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/competitions/_competition_form.html.erb:310]: Sorry I realised this only now but you're adding a field about `qualification_results_reason` which seems unrelated (and is also included further down in the same form!)

Perhaps a mistake from merging, please fix.

File: WcaOnRails/db/structure.sql
Comment (gregorbg) [inline WcaOnRails/db/structure.sql:749]: Unrelated change

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (gregorbg) [inline WcaOnRails/app/views/competitions/_competition_form.html.erb:309]: This is tricky because it uses a method on `competition` while filling out a form that modifies that same competition. I'd rather have the Yes/No dropdown be controlled by jQuery, much like `qualification_results_reason`.

(Perhaps you can even use the same jQuery `.on('change'` body for both? Cf. `assets/javascripts/competitions.js:30`

**Review by gregorbg — CHANGES_REQUESTED** [review]
Noticed some quirks while trying the feature out locally before merging 🙃

**Comment by gregorbg** [comment]
Kindly ping @timreyn. What's the status on this issue?

**Comment by timreyn** [comment]
Sorry, got busy with local competition stuff. I have a long weekend this weekend and should be able to resolve the remaining issues!

**Comment by gregorbg** [comment]
Cool! No rush, just wanted to check because the WST planning meeting for September is coming up.
Thanks for the quick update! :blush:

File: WcaOnRails/db/structure.sql
Comment (timreyn) [reply WcaOnRails/db/structure.sql:751]: Argh, every time I edit the db schema unrelated changes creep in...

Fixed.

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (timreyn) [reply WcaOnRails/app/views/competitions/_competition_form.html.erb]: Sounds good, done.

File: WcaOnRails/app/views/competitions/_competition_form.html.erb
Comment (timreyn) [reply WcaOnRails/app/views/competitions/_competition_form.html.erb]: Added to the edit below.

**Review by timreyn — COMMENTED** [review]
Looks like some extraneous files under webpacker/packs/generated_locales got added. Manually reverting these.

**Comment by gregorbg** [comment]
> Looks like some extraneous files under webpacker/packs/generated_locales got added. Manually reverting these.

Yes and No :sweat_smile: Between now and the time this PR was originally opened, we changed our JS i18n infrastructure. It used to write stuff to the `generated_locales` folder which we now renamed to `rails_translations` as part of the update process.
Arguably it was a bit unclever to un-ignore the old `generated_locales` folder because now older PRs like yours are trying to commit them :see_no_evil: Sorry!

**Review by gregorbg — CHANGES_REQUESTED** [review]
Code is all good now, but I hate to point out that `structure.sql` is pretty messed up again :see_no_evil:

Sorry, we have to endure this as long as there's still PHP portions of the code relying on our DB structure and DSL... Very very high on WST's radar to get rid of this!

**Comment by timreyn** [comment]
D'oh! Okay, should be fixed again.

**Review by gregorbg — APPROVED** [review]

**Pull request merged**
